test/ValuedCSP.lean

```diff
@@ -0,0 +1,173 @@
+    {Γ : ValuedCSP D C} {ι : Type*} {f : D → C}
+    (ok : ⟨1, Function.OfArity.uncurry f⟩ ∈ Γ) (i : ι) : Γ.Term ι :=
+  ⟨1, Function.OfArity.uncurry f, ok, ![i]⟩
+
+def ValuedCSP.binaryTerm {D C : Type} [OrderedAddCommMonoid C]
+    {Γ : ValuedCSP D C} {ι : Type*} {f : D → D → C}
+    (ok : ⟨2, Function.OfArity.uncurry f⟩ ∈ Γ) (i j : ι) : Γ.Term ι :=
+  ⟨2, Function.OfArity.uncurry f, ok, ![i, j]⟩
+
+-- ## Example: minimize `|x| + |y|` where `x` and `y` are rational numbers
+
+private def absRat : (Fin 1 → ℚ) → ℚ := Function.OfArity.uncurry abs
+
+private def exampleAbs : Σ (n : ℕ), (Fin n → ℚ) → ℚ := ⟨1, absRat⟩
+
+private def exampleFiniteValuedCSP : ValuedCSP ℚ ℚ := {exampleAbs}
+
+private lemma abs_in : ⟨1, absRat⟩ ∈ exampleFiniteValuedCSP := rfl
+
+private def exampleFiniteValuedInstance : exampleFiniteValuedCSP.Instance (Fin 2) :=
+  Multiset.ofList [ValuedCSP.unaryTerm abs_in 0, ValuedCSP.unaryTerm abs_in 1]
+
+example : exampleFiniteValuedInstance.IsOptimumSolution ![(0 : ℚ), (0 : ℚ)] := by
+  unfold ValuedCSP.Instance.IsOptimumSolution
+  unfold exampleFiniteValuedCSP
+  intro s
+  convert_to 0 ≤ exampleFiniteValuedInstance.evalSolution s
+  rw [ValuedCSP.Instance.evalSolution, exampleFiniteValuedInstance]
+  convert_to 0 ≤ |s 0| + |s 1|
+  · simp [ValuedCSP.unaryTerm, ValuedCSP.Term.evalSolution, Function.OfArity.uncurry]
+  positivity
+
+-- ## Example: `B ≠ A ≠ C ≠ D ≠ B ≠ C` with three available labels (i.e., 3-coloring of K₄⁻)
+
+private def Bool_add_le_add_left (a b : Bool) :
+  (a ≤ b) → ∀ (c : Bool), ((c || a) ≤ (c || b)) :=
+by
+  intro hab c
+  cases a <;> cases b <;> cases c <;> trivial
+
+-- For simpler implementation, we treat `false` as "satisfied" and `true` as "wrong" here.
+instance crispCodomain : LinearOrderedAddCommMonoid Bool where
+  __ := Bool.linearOrder
+  add (a b : Bool) := a || b
+  add_assoc := Bool.or_assoc
+  zero := false
+  zero_add (_ : Bool) := rfl
+  add_zero := Bool.or_false
+  add_comm := Bool.or_comm
+  add_le_add_left := Bool_add_le_add_left
+
+private def beqBool : (Fin 2 → Fin 3) → Bool := Function.OfArity.uncurry BEq.beq
+
+private def exampleEquality : Σ (n : ℕ), (Fin n → Fin 3) → Bool := ⟨2, beqBool⟩
+
+private def exampleCrispCsp : ValuedCSP (Fin 3) Bool := {exampleEquality}
+
+private lemma beq_in : ⟨2, beqBool⟩ ∈ exampleCrispCsp := rfl
+
+private def exampleTermAB : exampleCrispCsp.Term (Fin 4) :=
+  ValuedCSP.binaryTerm beq_in 0 1
+
+private def exampleTermBC : exampleCrispCsp.Term (Fin 4) :=
+  ValuedCSP.binaryTerm beq_in 1 2
+
+private def exampleTermCA : exampleCrispCsp.Term (Fin 4) :=
+  ValuedCSP.binaryTerm beq_in 2 0
+
+private def exampleTermBD : exampleCrispCsp.Term (Fin 4) :=
+  ValuedCSP.binaryTerm beq_in 1 3
+
+private def exampleTermCD : exampleCrispCsp.Term (Fin 4) :=
+  ValuedCSP.binaryTerm beq_in 2 3
+
+private def exampleCrispCspInstance : exampleCrispCsp.Instance (Fin 4) :=
+  Multiset.ofList [exampleTermAB, exampleTermBC, exampleTermCA, exampleTermBD, exampleTermCD]
+
```

Comment:
Perhaps a comment explaining the example below. I am having trouble understanding why these 12 assignments were chosen as examples.

Reply:
Would any of the given documentation comments in ASCII art help you?

Reply:
Ah yes, the ASCII art is extremely helpful, thanks

Reply:
I think I like the last format best

Reply:
I will add them to all examples. As normal comments (not documentation comments), because autogenerating docs would probably make mess of it.